File: src/engine/kb.ts
// kb.ts — Knowledge Base: TELL / ASK API

import type { Clause, Literal } from './cnf';
import { clauseKey } from './cnf';
import { resolve } from './resolution';

export interface KnowledgeBase {
  /** All clauses currently in the KB. */
  clauses: Clause[];
  /** Human-readable log of what was told to the KB. */
  tellLog: string[];
}

export interface AskResult {
  proved: boolean;
  inferenceSteps: number;
  log: string[];
}

/** Create an empty Knowledge Base. */
export function createKB(): KnowledgeBase {
  return { clauses: [], tellLog: [] };
}

/**
 * Add new clauses to the KB, deduplicating by canonical clause key.
 * Returns a new KB (immutable update).
 */
export function tellKB(kb: KnowledgeBase, newClauses: Clause[], logMsg: string): KnowledgeBase {
  const existingKeys = new Set(kb.clauses.map(clauseKey));
  const unique = newClauses.filter(c => !existingKeys.has(clauseKey(c)));
  return {
    clauses: [...kb.clauses, ...unique],
    tellLog: [...kb.tellLog, logMsg],
  };
}

/**
 * ASK the KB whether a given literal can be proved via resolution refutation.
 *
 * To prove "¬P_2_1" (cell 2,1 has no pit):
 *   - Add the clause [P_2_1] (negation of goal) to a copy of the KB
 *   - Run resolution — if empty clause derived → proved
 */
export function askKB(kb: KnowledgeBase, goal: Literal): AskResult {
  return resolve(kb.clauses, goal);
}

File: src/python_engine/kb.py
from typing import List, Dict, Any
from .cnf import Clause, Literal, clause_key
from .resolution import AskResult, resolve

class KnowledgeBase:
    def __init__(self, clauses: List[Clause] = None, tell_log: List[str] = None):
        self.clauses = clauses if clauses is not None else []
        self.tell_log = tell_log if tell_log is not None else []

def create_kb() -> KnowledgeBase:
    return KnowledgeBase()

def tell_kb(kb: KnowledgeBase, new_clauses: List[Clause], log_msg: str) -> KnowledgeBase:
    existing_keys = {clause_key(c) for c in kb.clauses}
    unique = [c for c in new_clauses if clause_key(c) not in existing_keys]
    
    return KnowledgeBase(
        clauses=kb.clauses + unique,
        tell_log=kb.tell_log + [log_msg]
    )

def ask_kb(kb: KnowledgeBase, goal: Literal) -> AskResult:
    return resolve(kb.clauses, goal)
